aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -567,5 +568,270 @@ Tensor& index_add_cuda_(Tensor & self, int64_t dim, const Tensor & index, const
+      IndexType dstOffset =
+        cuda::detail::IndexToOffset<T, IndexType, DstDim>::get(linearIndex, dst);
+      dstOffset += dstIndex * dst.strides[dstSelectDim];
+
+      IndexType srcOffset =
+        cuda::detail::IndexToOffset<T, IndexType, SrcDim>::get(linearIndex, src);
+      srcOffset += srcIndex * src.strides[srcSelectDim];
+
+      dst.data[dstOffset] = src.data[srcOffset];
+    }
+  }
+}
+
+// We prefer this kernel to balance parallelism across index points,
+// if there are a large number of indices.
+// This kernel in fact works for all choices of problem size, but if
+// the number of indices chosen is small, then the
+// indexSelectSmallIndex kernel is a better choice to reduce memory
+// accesses.
+template <typename T, typename IndexType, int DstDim, int SrcDim, int IdxDim,
+          bool IndexIsMajor>
+__global__ void indexSelectLargeIndex(cuda::detail::TensorInfo<T, IndexType> dst,
+                                      cuda::detail::TensorInfo<T, IndexType> src,
+                                      cuda::detail::TensorInfo<int64_t, IndexType> indices,
+                                      int dstSelectDim,
+                                      int srcSelectDim,
+                                      IndexType totalSize,
+                                      IndexType innerSize,
+                                      int64_t srcSelectDimSize) {
+  // We stride over the output including the indexed dimension
+  // (totalSize), and calculate the destination index point based on that
+  for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < totalSize;
+       linearIndex += gridDim.x * blockDim.x) {
+    IndexType dstIndex, elementInSlice;
+    if (IndexIsMajor) {
+      dstIndex = linearIndex / innerSize;
+      elementInSlice = linearIndex % innerSize;
+    }
+    else {
+      elementInSlice = linearIndex / innerSize;
+      dstIndex = linearIndex % innerSize;
+    }
+
+    IndexType srcIndex =
+      indices.data[cuda::detail::IndexToOffset<int64_t, IndexType, IdxDim>::get(dstIndex, indices)];
+    CUDA_KERNEL_ASSERT(srcIndex < srcSelectDimSize);
+
+    IndexType dstOffset =
+      cuda::detail::IndexToOffset<T, IndexType, DstDim>::get(elementInSlice, dst);
+    dstOffset += dstIndex * dst.strides[dstSelectDim];
+
+    IndexType srcOffset =
+      cuda::detail::IndexToOffset<T, IndexType, SrcDim>::get(elementInSlice, src);
+    srcOffset += srcIndex * src.strides[srcSelectDim];
+
+    dst.data[dstOffset] = src.data[srcOffset];
+  }
+}
+
+namespace {
+
+// When using a 0-dim scalar tensor, we need the legacy (THC) semantics of
+// TensorInfo: Pretend that the scalar tensor is in fact a one-element vector.
+template <typename T, typename IndexType>
+cuda::detail::TensorInfo<T, IndexType>
+tensorInfoLegacyIfScalar(cuda::detail::TensorInfo<T, IndexType> ti) {
+  if (ti.dims == 0) {
+    ti.dims = 1;
+    ti.sizes[0] = 1;
+    ti.strides[0] = 1;
+  }
+  return ti;
+}
```

Comment:
As an alternative I have introduced this function to be wrapped around getTensorInfo. Please let me know if you don't think that's a good approach and I will be happy to revert.

Reply:
Sounds like a great solution.